### PR TITLE
Enable IPv6 Support for Provisioning 

### PR DESCRIPTION
**Pull request by hardys**
By setting PROVISIONING_IPV6=true we configure the provisioning
bridge with an ipv6 address only, to enable further testing
of single-stack ipv6 provisioning

Related PR's
- [x] https://github.com/metal3-io/baremetal-operator/pull/324
- [x] https://github.com/metal3-io/metal3-dev-env/pull/104
- [x] https://github.com/metal3-io/ironic-image/pull/107
- [x] https://github.com/metal3-io/ironic-inspector-image/pull/42
- [x] https://github.com/metal3-io/ironic-ipa-downloader/pull/11
- [x] https://github.com/metal3-io/ironic-ipa-downloader/pull/10
- [x] https://review.opendev.org/#/c/695546/
- [x] https://review.opendev.org/#/c/692110/
- [x] https://github.com/metal3-io/ironic-inspector-image/pull/43

### CODE REVIEW

**Comment by stbenjam** [comment]
Status:
 - BMO can deploy with custom config map
 - Sample config map with IPv6 config
 - httpd can now start and listens on IPv6 address with correct inspector URL configured

File: lib/common.sh [hunk 1]
```diff
@@ -67,7 +67,7 @@ export SUSHY_TOOLS_IMAGE=${SUSHY_TOOLS_IMAGE:-"quay.io/metal3-io/sushy-tools"}
 
 # Ironic vars
 export IPA_DOWNLOADER_IMAGE=${IPA_DOWNLOADER_IMAGE:-"quay.io/metal3-io/ironic-ipa-downloader:master"}
-export IRONIC_IMAGE=${IRONIC_IMAGE:-"quay.io/metal3-io/ironic:master"}
+export IRONIC_IMAGE=${IRONIC_IMAGE:-"quay.io/stbenjam/ironic-image:latest"} # FIXME: Built from 'ipv6' branch on my fork
```
Comment (hardys) [inline lib/common.sh]: Tangential to the PR but I wonder if we could refactor and reuse the local image stuff @derekhiggins added to dev-scripts to make testing with custom image branches easier

**Comment by stbenjam** [comment]
I've got the kustomize version of things working, and got the basics for local image support up.

Ironic API is also up and responding on IPv6 now.

**Comment by stbenjam** [comment]
The way I've been testing this is

```
export IRONIC_LOCAL_IMAGE=file:///home/notstack/ironic-image
export IRONIC_INSPECTOR_LOCAL_IMAGE=file:///home/notstack/ironic-inspector-image
```

With the PR's linked in the description checked out in those paths.

File: ipv6/kustomization.yaml [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+  literals:
+  - PROVISIONING_IP=fd2e:6f44:5dd8:b856::2
+  - DEPLOY_KERNEL_URL=http://[fd2e:6f44:5dd8:b856::2]:6180/images/ironic-python-agent.kernel
+  - DEPLOY_RAMDISK_URL=http://[fd2e:6f44:5dd8:b856::2]:6180/images/ironic-python-agent.initramfs
+  - IRONIC_ENDPOINT=http://[fd2e:6f44:5dd8:b856::2]:6385/v1/
+  - IRONIC_INSPECTOR_ENDPOINT=http://[fd2e:6f44:5dd8:b856::2]:5050/v1/
+  - CACHEURL=http://[fd2e:6f44:5dd8:b856::1]/images
+  name: ironic-bmo-configmap
+resources:
+- ../../go/src/github.com/metal3-io/baremetal-operator/deploy
```
Comment (stbenjam) [inline ipv6/kustomization.yaml]: This assumes dev-scripts is checked out in $HOME, unfortunately kustomize doesn't have a way to change the base path easily, might need to be a sed thing or generate this whole file in a script.

File: 01_prepare_host.sh [hunk 3]
```diff
@@ -55,7 +55,7 @@ podman)
 esac
 
 # Pull images we'll need
-for IMAGE_VAR in IRONIC_IMAGE IPA_DOWNLOADER_IMAGE VBMC_IMAGE SUSHY_TOOLS_IMAGE; do
+for IMAGE_VAR in IRONIC_IMAGE IPA_DOWNLOADER_IMAGE VBMC_IMAGE SUSHY_TOOLS_IMAGE DOCKER_REGISTRY_IMAGE; do
```
Comment (hardys) [inline 01_prepare_host.sh]: We need to add the logic from dev-scripts `01_install_requirements.sh` to start the docker-distribution container, as this only pulls it I think?

File: 02_configure_host.sh [hunk 4]
```diff
@@ -123,6 +123,11 @@ for port in 6180 5050 6385 9999; do
     fi
 done
 
+# Add firewall rules to allow access to registry
```
Comment (hardys) [inline 02_configure_host.sh]: We could perhaps rebase this on your firewall PR since it's needed for testing on rhel hosts and looks close to landing?

File: 01_prepare_host.sh (hunk 3 above)
Comment (hardys) [reply 01_prepare_host.sh]: Oh nvm I see it's below...

File: 03_launch_mgmt_cluster.sh [hunk 5]
```diff
@@ -98,7 +98,38 @@ function clone_repos() {
     popd
 }
 
+function build_local_images() {
```
Comment (hardys) [inline 03_launch_mgmt_cluster.sh]: Not in this pr but it'd be nice to figure out a way to de-duplicate this code between dev-scripts and metal3-dev-env

File: ipv6/kustomization.yaml (hunk 2 above)
Comment (hardys) [reply ipv6/kustomization.yaml]: resources can't take an absolute path?

Comment (hardys) [reply ipv6/kustomization.yaml]: Oh I see even with an absolute path we'll need a variable to say where the `$GOPATH` is

File: ipv6/kustomization.yaml [hunk 6]
```diff
@@ -1,9 +1,10 @@
 apiVersion: kustomize.config.k8s.io/v1beta1
 kind: Kustomization
 configMapGenerator:
-- behavior: replace
+- behavior: merge
   literals:
   - PROVISIONING_IP=fd2e:6f44:5dd8:b856::2
+  - DHCP_RANGE=fd2e:6f44:5dd8:b856::3,fd2e:6f44:5dd8:b856::ff
```
Comment (hardys) [inline ipv6/kustomization.yaml]: In the v4 sample we have a range of `.10` to `.100` - we should perhaps do the same here (particularly the start range to avoid potential conflict with the cluster-hosted ironic in the openshift case as that uses `.3` atm?)

Comment (stbenjam) [reply ipv6/kustomization.yaml]: Sure, so `::A` to just `::64`?

Comment (hardys) [reply ipv6/kustomization.yaml]: yup, just so we know there's no chance of the DHCP leases overlapping with other static addresses at the start of the subnet range.

Comment (hardys) [reply ipv6/kustomization.yaml]: I suppose we could leave the upper bound at the subnet end, and just ensure the start is offset from the beginning?

Comment (derekhiggins) [reply ipv6/kustomization.yaml]: I like the idea of giving it a default that will be big enough for production and avoids the start of the range, something like a /96
DHCP_RANGE=fd2e:6f44:5dd8:b856:0:1:0:0,fd2e:6f44:5dd8:b856:0:1:fff:fff

**Comment by derekhiggins** [comment]
Changes required in the IPA image to support IPv6
Bump up ipa-ip-lookup-attempts to 6 - https://review.opendev.org/#/c/695546/ (merged in master)
Add IPv6 support in dhcp-all-interfaces - https://review.opendev.org/#/c/692110/

**Comment by hardys** [comment]
> Changes required in the IPA image to support IPv6
> Bump up ipa-ip-lookup-attempts to 6 - https://review.opendev.org/#/c/695546/ (merged in master)
> Add IPv6 support in dhcp-all-interfaces - https://review.opendev.org/#/c/692110/

These both landed, as did the ironic-image changes - @stbenjam would you like to rebase/squash the series now so we can start a final review/test pass?

I'm thinking we could land this iteration then work on switching the BMC ips to v6 as a follow-up?

**Comment by stbenjam** [comment]
> > Changes required in the IPA image to support IPv6
> > Bump up ipa-ip-lookup-attempts to 6 - [review.opendev.org/#/c/695546](https://review.opendev.org/#/c/695546/) (merged in master)
> > Add IPv6 support in dhcp-all-interfaces - [review.opendev.org/#/c/692110](https://review.opendev.org/#/c/692110/)
> 
> These both landed, as did the ironic-image changes - @stbenjam would you like to rebase/squash the series now so we can start a final review/test pass?
> 
> I'm thinking we could land this iteration then work on switching the BMC ips to v6 as a follow-up?

That works for me. We need https://github.com/metal3-io/ironic-inspector-image/pull/43 to land first, though.  I had fixed it locally and never pushed the changes, which is why my image on quay.io was working but the one in metal3-io didn't.

File: scripts/v1alpha1/create_machine.sh [hunk 7]
```diff
@@ -9,10 +10,17 @@ V1ALPHA1_SCRIPTS_PATH="$(dirname "$(readlink -f "${0}")")/"
 USER_DATA_SCRIPT_NAME="user_data.sh"
 SCRIPT_PATH="${V1ALPHA1_SCRIPTS_PATH}""${USER_DATA_SCRIPT_NAME}"
 
+if [[ "${PROVISIONING_IPV6}" == "true" ]]; then
+  PROVISIONING_HOST="[fd2e:6f44:5dd8:b856::1]"
+else
+  PROVISIONING_HOST=172.22.0.1
+fi
```
Comment (hardys) [inline scripts/v1alpha1/create_machine.sh]: Can we just use `PROVISIONING_IP` from `lib/common.sh` here?

File: scripts/v1alpha1/create_machine.sh [hunk 8]
```diff
@@ -9,10 +10,17 @@ V1ALPHA1_SCRIPTS_PATH="$(dirname "$(readlink -f "${0}")")/"
...
+
 MACHINE_NAME=$1
 IMAGE_NAME=${2:-${IMAGE_NAME}}
-IMAGE_URL=http://172.22.0.1/images/${IMAGE_NAME}
-IMAGE_CHECKSUM=http://172.22.0.1/images/${IMAGE_NAME}.md5sum
+
+IMAGE_URL=http://$PROVISIONING_HOST/images/${IMAGE_NAME}
+IMAGE_CHECKSUM=http://$PROVISIONING_HOST/images/${IMAGE_NAME}.md5sum
```
Comment (hardys) [inline scripts/v1alpha1/create_machine.sh]: Likewise, if we reorder the definitions in lib/common./sh this can be removed I think?

Comment (stbenjam) [reply scripts/v1alpha1/create_machine.sh]: The script lets you override the IMAGE_NAME as an argument, so I think this has to stay

File: scripts/v1alpha1/create_machine.sh (hunk 7 above)
Comment (stbenjam) [reply scripts/v1alpha1/create_machine.sh]: Ah yes, thanks, done

**Comment by hardys** [comment]
@maelk Hi - do you have access to any logs which may help figure out why the integration test gets stuck like:

```
FAIL - node-0 Baremetalhost introspecting completed
       expected inspecting to be in ready available
```

I'm not clear atm if that's a false negative or if the PR is really breaking something.

**Comment by maelk** [comment]
Hi! Sorry for the delay, I will look into this. The integration tests are broken, Baremetal-operator crashes and everything get stuck. I don't know yet why it crashes though

**Comment by maelk** [comment]
The tests are passing on master, this would need a rebase to pass the tests

**Comment by hardys** [comment]
@stbenjam  I tested locally and I think the CI failures are real - I see the nodes fail to start with a libvirt error starting the domain - it can't find `/usr/share/OVMF/OVMF_VARS.fd`

Does that sound familiar?

**Comment by stbenjam** [comment]
That'd be the OVMF firmware when doing UEFI boot, are you testing ipv6?

**Comment by stbenjam** [comment]
Looks like we forgot to explicitly installed edk2-ovmf (EL) and ovmf (Ubuntu) packages, I've added them to the Ansible roles.

File: vm-setup/roles/packages_installation/defaults/main.yml [hunk 9]
```diff
@@ -1,21 +1,22 @@
...
-     - crudini 
-     - curl 
-     - dnsmasq 
-     - nmap 
-     - patch 
-     - psmisc 
+     - crudini
+     - curl
+     - dnsmasq
+     - ovmf
```
Comment (stbenjam) [inline vm-setup/roles/packages_installation/defaults/main.yml]: This is added to the Ubuntu list, and `edk2-ovmf` below.  My editor removed the trailing whitespace, I can undo that if we want to keep it.

File: vm-setup/roles/packages_installation/defaults/main.yml [hunk 10]
```diff
@@ -26,89 +27,90 @@ packages:
...
-         - python-pip 
-         - python-requests 
-         - python-setuptools 
-         - vim-enhanced 
+         - crudini
+         - curl
+         - dnsmasq
+         - NetworkManager
+         - nmap
+         - patch
```
Comment (hardys) [inline vm-setup/roles/packages_installation/defaults/main.yml]: We're still missing the package in the centos7 case, I did yum install OVMF which appears to contain the right files

Comment (hardys) [reply vm-setup/roles/packages_installation/defaults/main.yml]: It's a bit hard to follow since there are unrelated sort/linefeed changes here too.

**Comment by hardys** [comment]
Thanks, this worked for me locally with ipv4 so lets hope CI agrees :)

**Comment by stbenjam** [comment]
Tests are green!!

**Comment by hardys** [comment]
Good to see the CI pass now, and this works fine locally on centos7 without PROVISIONING_IPV6 set, but when I enable ipv6 we fail to pxeboot the ramdisk.

I wonder if that's a centos7 vs el8 thing, I'll re-test on el8 to confirm tomorrow.

**Comment by hardys** [comment]
Ok I re-tested this on a freshly installed RHEL8 host and it worked, although the pxe boot of node-1 hung and only worked after a manual force-reset of the VM.

@stbenjam I think we should go ahead and merge this, then iterate on any remaining issues, but do you think we should either add a comment in the example config that this won't work on centos7, or raise an issue to track that?

I suspect we may want to add centos8 support and just say this is unsupported on centos7, otherwise we'll have to figure out what the problems are with the roms on centos7.

**Comment by stbenjam** [comment]
> Ok I re-tested this on a freshly installed RHEL8 host and it worked, although the pxe boot of node-1 hung and only worked after a manual force-reset of the VM.
> 
> @stbenjam I think we should go ahead and merge this, then iterate on any remaining issues, but do you think we should either add a comment in the example config that this won't work on centos7, or raise an issue to track that?
> 
> I suspect we may want to add centos8 support and just say this is unsupported on centos7, otherwise we'll have to figure out what the problems are with the roms on centos7.

There weren't any CentOS 8 cloud images when I looked. Today it looks like there are vagrant images in http://cloud.centos.org/centos/8/, so maybe they're getting closer to getting them all built.

The reason CentOS 7 doesn't work is because the cloud images are BIOS only.  FCOS and Ubuntu both have dual BIOS/UEFI images. Hopefully when CentOS 8 images come out they will be dual boot.

**Comment by hardys** [comment]
Ok I raised https://github.com/metal3-io/metal3-dev-env/issues/149 to track getting some support for centos w/ipv6 - lets merge this and iterate on that as a follow-up.
